### ISSUE TITLE
[RFC] Combining static thresholds and streaming ML modeling for anomaly detection

### ISSUE THREAD

**Issue by kaituo** [opensearch-project/anomaly-detection#563]
**What problems are you trying to solve?**

Machine learning (ML) based anomaly detection outputs rare data patterns via statistical analysis of past data. However, no machine learning (ML) model is perfect. For example, an anomaly detection model may rightly find a spike but miss the mark in determining the extent of spikes to trigger an alert. Such misclassifications falsely alert customers to non-issues, or unnecessarily miss critical events. It would be nice if a user could equip an AD plugin with a safe threshold to trigger alarms when AD detects a changed pattern. Currently, AD cannot incorporate rules in its workflow.

**What are you proposing?**

Combining ML-based anomaly detection and domain-specific rules improve prediction performance. OpenSearch customers will be able to tell the model what it should do besides the models' findings. This information allows the customer to tune anomaly detection performance to their specific needs on their data.

**What is the user experience going to be?**

AD allows users to specify rules deriving an anomaly before or after the detector runs. We link rules via AND/OR. That feature X's value should be at least y before being considered an anomaly describes an AND linkage between the threshold condition (X > y) and anomaly condition (anomaly grade > 0). Reporting an anomaly as long as feature X's value is larger than Y regardless of what the model says describes an OR linkage between the same two conditions.

**Why should it be built? Any reason not to?**

Some community users felt the necessity of combining domain-specific knowledge and ML models. One user anecdotally revealed their interest in marrying static thresholds with ML models. Another [forum user](https://discuss.opendistrocommunity.dev/t/using-anomaly-detection-to-detect-sudden-increases-or-decreases-in-the-amount-of-logs-received-for-an-index/8628) indicates that the ML model reports too many anomalies and wants to specify the minimum percentage of changes for an anomaly. Another [github issue](https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/388) requests to change the threshold condition after ML models have learned the threshold range.

**What will it take to execute?**

When a user can provide rules deriving an anomaly, we can combine the rules and ML models to achieve combined benefits. The solution is equivalent to adding conjuncts (if the metric value is at least the threshold before being considered an anomaly) or disjuncts (if the metric value exceeds the threshold, then mark the observation as anomalous) to the original anomaly grade > 0 condition. There are many kinds of rules. One rule is a value outside the range. For example, consider a feature value larger than x. Another rule is missing data. For instance, 90% of past x intervals have missing data. Further, we can leverage temporal information in a time series rather than their instantaneous values alone. Users can configure the alert to happen if something happened "just once," "for the last y readings," or "for z of the last y readings"? The alert can also occur based on change percentage. For example, the change from one point to the next has to be at least x%.

**What are remaining open questions?**

How to implement a low-latency and high-volume rule evaluation when the rule has temporal information since there can be repeated queries?

**Request for comments**

We are looking for all forms of feedback! Some questions we would like to get your input on include

* What are your use cases of static thresholds in anomaly detection?
* Do you have additional feedback beyond what is listed in this RFC?

**Comment by xinlamzn** [comment]
This sounds like the post-processing of AD results, focusing on when should we notify customers on the anomaly. We may build more on Alerting strategy, including suppression, aggregation, etc.

**Comment by kaituo** [comment]
Yes, it is post-processing of AD results. We may be able to do these in alerting. But doing this inside AD can improve the easiness of use. They can directly view filtered anomalies in AD without creating a separate dashboard. They don't have to write complicated queries in alerting.
